File: c-password/password-generator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h> // Include the time header

#define LOWERCASE_CHARS "abcdefghijklmnopqrstuvwxyz"
#define UPPERCASE_CHARS "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
#define NUMBERS "0123456789"
#define SYMBOLS "!@#$%^&*_=+-/"

typedef struct {
    int chars_count;
    int caps_count;
    int numbers_count;
    int symbols_count;
} PasswordGenerator;

void initializeGenerator(PasswordGenerator *generator, int chars_count, int caps_count, int numbers_count, int symbols_count) {
    generator->chars_count = chars_count > 0 ? chars_count : 0;
    generator->caps_count = caps_count > 0 ? caps_count : 0;
    generator->numbers_count = numbers_count > 0 ? numbers_count : 0;
    generator->symbols_count = symbols_count > 0 ? symbols_count : 0;
}

void generatePasswordSpecific(PasswordGenerator *generator, char *password) {
    sprintf(password, "%.*s%.*s%.*s%.*s",
            generator->chars_count, LOWERCASE_CHARS,
            generator->caps_count, UPPERCASE_CHARS,
            generator->numbers_count, NUMBERS,
            generator->symbols_count, SYMBOLS);
    // Shuffle the password
    for (int i = strlen(password) - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        char temp = password[i];
        password[i] = password[j];
        password[j] = temp;
    }
}

void generatePassword(PasswordGenerator *generator, char *password) {
    generatePasswordSpecific(generator, password);
}

void generatePasswordWithoutSymbols(PasswordGenerator *generator, char *password) {
    generator->symbols_count = 0;
    generatePasswordSpecific(generator, password);
}

void generatePasswordWithNumbersOnly(PasswordGenerator *generator, char *password) {
    generator->chars_count = 0;
    generator->caps_count = 0;
    generator->symbols_count = 0;
    generator->numbers_count = 6; // Set numbers_count to 6 for exactly 6 digits
    generatePasswordSpecific(generator, password);
}

int main() {
    srand((unsigned int)time(NULL));

    PasswordGenerator pwdGenerator;
    initializeGenerator(&pwdGenerator, 4, 2, 2, 1);

    char generatedPassword[20];  // Adjust the size based on your needs

    printf("Password with characters, caps, numbers, and symbols:\n");
    generatePassword(&pwdGenerator, generatedPassword);
    printf("%s\n", generatedPassword);

    printf("Another password with the same set:\n");
    generatePassword(&pwdGenerator, generatedPassword);
    printf("%s\n", generatedPassword);

    printf("Password with characters, caps, numbers, without symbols:\n");
    generatePasswordWithoutSymbols(&pwdGenerator, generatedPassword);
    printf("%s\n", generatedPassword);

    printf("Password with numbers only (exactly 6 digits):\n");
    generatePasswordWithNumbersOnly(&pwdGenerator, generatedPassword);
    printf("%s\n", generatedPassword);

    return 0;
}

File: java-password/password-generator.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class PasswordGenerator {
    private static final String LOWERCASE_CHARS = "abcdefghijklmnopqrstuvwxyz";
    private static final String UPPERCASE_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final String NUMBERS = "0123456789";
    private static final String SYMBOLS = "!@#$%^&*_=+-/";

    private int charsCount;
    private int capsCount;
    private int numbersCount;
    private int symbolsCount;

    public PasswordGenerator(int charsCount, int capsCount, int numbersCount, int symbolsCount) {
        this.charsCount = Math.max(charsCount, 0);
        this.capsCount = Math.max(capsCount, 0);
        this.numbersCount = Math.max(numbersCount, 0);
        this.symbolsCount = Math.max(symbolsCount, 0);
    }

    public String generatePassword() {
        return generatePasswordSpecific(charsCount, capsCount, numbersCount, symbolsCount);
    }

    public String generatePasswordWithoutSymbols() {
        return generatePasswordSpecific(charsCount, capsCount, numbersCount, 0);
    }

    public String generatePasswordWithNumbersOnly() {
        return getRandomString(NUMBERS, 6);
    }

    private String generatePasswordSpecific(int charsCount, int capsCount, int numbersCount, int symbolsCount) {
        StringBuilder password = new StringBuilder();
        password.append(getRandomString(LOWERCASE_CHARS, charsCount))
                .append(getRandomString(UPPERCASE_CHARS, capsCount))
                .append(getRandomString(NUMBERS, numbersCount))
                .append(getRandomString(SYMBOLS, symbolsCount));

        return shuffle(password.toString());
    }

    private String getRandomString(String charSet, int count) {
        List<Character> characters = new ArrayList<>();
        for (char c : charSet.toCharArray()) {
            characters.add(c);
        }
        Collections.shuffle(characters);

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < count; i++) {
            result.append(characters.get(i));
        }

        return result.toString();
    }

    private String shuffle(String source) {
        List<Character> characters = new ArrayList<>();
        for (char c : source.toCharArray()) {
            characters.add(c);
        }
        Collections.shuffle(characters);

        StringBuilder result = new StringBuilder();
        for (char c : characters) {
            result.append(c);
        }

        return result.toString();
    }

    public static void main(String[] args) {
        PasswordGenerator pwdGenerator = new PasswordGenerator(4, 2, 2, 1);

        System.out.println("Password with characters, caps, numbers, and symbols:");
        System.out.println(pwdGenerator.generatePassword());

        System.out.println("Another password with the same set:");
        System.out.println(pwdGenerator.generatePassword());

        System.out.println("Password with characters, caps, numbers, without symbols:");
        System.out.println(pwdGenerator.generatePasswordWithoutSymbols());

        System.out.println("Password with numbers only (exactly 6 digits):");
        System.out.println(pwdGenerator.generatePasswordWithNumbersOnly());
    }
}
